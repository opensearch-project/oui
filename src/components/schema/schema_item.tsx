/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React, { FunctionComponent } from 'react';
import { CommonProps } from '../common';
import classNames from 'classnames';
import { OuiButtonIcon, OuiButtonIconPropsForButton } from '../button';
import { OuiToolTip, OuiToolTipProps } from '../tool_tip';
import { IconType } from '../icon';
import { OuiToken } from '../token';

export interface Action extends OuiButtonIconPropsForButton {
  tooltip?: Omit<OuiToolTipProps, 'children'>;
}

/**
 * Props for the OuiSchemaItem component.
 * @public
 */
export type OuiSchemaItemProps = CommonProps & {
  /** The icon type to display. */
  iconType?: IconType;
  /** The label to display. */
  label: string;
  /** An array of actions to display. Limit 2 */
  actions?: Action[];
  /** Whether the item is compressed. */
  compressed?: boolean;
  /** Whether the item is displayed with a panel. */
  withPanel?: boolean;
};

/**
 * A component that displays a schema item.
 * @public
 */
export const OuiSchemaItem: FunctionComponent<OuiSchemaItemProps> = ({
  className,
  iconType,
  label,
  actions = [],
  compressed,
  withPanel,
  ...rest
}) => {
  const classes = classNames(
    'ouiSchemaItem',
    {
      'ouiSchemaItem--compressed': compressed,
      'ouiSchemaItem--withPanel': withPanel,
    },
    className
  );

  const filteredActions = actions.slice(0, 2);

  return (
    <div className={classes} {...rest}>
      {iconType && (
        <OuiToken
          className="ouiSchemaItem__icon"
          iconType={iconType}
          size={compressed ? 'xs' : 's'}
        />
      )}
      <div className="ouiSchemaItem__label">{label}</div>
      <div className="ouiSchemaItem__actions">
        {filteredActions.map(({ tooltip, ...action }, index) =>
          tooltip ? (
            <OuiToolTip key={index} {...tooltip}>
              <OuiButtonIcon {...action} size={compressed ? 'xs' : 's'} />
            </OuiToolTip>
          ) : (
            <OuiButtonIcon
              key={index}
              {...action}
              size={compressed ? 'xs' : 's'}
            />
          )
        )}
      </div>
    </div>
  );
};
