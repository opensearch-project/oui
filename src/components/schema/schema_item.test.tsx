/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../test/required_props';

import { OuiSchemaItem } from './schema_item';

describe('OuiSchema', () => {
  test('is rendered', () => {
    const component = render(<OuiSchemaItem {...requiredProps} label="Test" />);

    expect(component).toMatchSnapshot();
  });

  test('it only renders at most 2 actions', () => {
    const actions = Array(3)
      .fill(null)
      .map(() => ({
        iconType: 'trash',
        onClick: () => {},
      }));
    const component = render(
      <OuiSchemaItem {...requiredProps} label="Test" actions={actions} />
    );
    expect(component.find('.ouiSchemaItem__actions').children().length).toBe(2);
  });

  test('it renders compressed correctly', () => {
    const component = render(
      <OuiSchemaItem {...requiredProps} label="Test" compressed />
    );
    expect(component).toMatchSnapshot();
  });
});
