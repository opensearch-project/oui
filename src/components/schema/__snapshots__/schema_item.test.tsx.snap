// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiSchema is rendered 1`] = `
<div
  aria-label="aria-label"
  class="ouiSchemaItem testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    class="ouiSchemaItem__label"
  >
    Test
  </div>
  <div
    class="ouiSchemaItem__actions"
  />
</div>
`;

exports[`OuiSchema it renders compressed correctly 1`] = `
<div
  aria-label="aria-label"
  class="ouiSchemaItem ouiSchemaItem--compressed testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    class="ouiSchemaItem__label"
  >
    Test
  </div>
  <div
    class="ouiSchemaItem__actions"
  />
</div>
`;
